lib/galaxy/objectstore/cloud.py

```diff
@@ -0,0 +1,565 @@
+        return rel_path
+
+    def _get_cache_path(self, rel_path):
+        return os.path.abspath(os.path.join(self.staging_path, rel_path))
+
+    def _get_transfer_progress(self):
+        return self.transfer_progress
+
+    def _get_size_in_cloud(self, rel_path):
+        try:
+            obj = self.bucket.get(rel_path)
+            if obj:
+                return obj.size
+        except Exception:
+            log.exception("Could not get size of key '%s' from S3", rel_path)
+            return -1
+
+    def _key_exists(self, rel_path):
+        exists = False
+        try:
+            # A hackish way of testing if the rel_path is a folder vs a file
+            is_dir = rel_path[-1] == '/'
+            if is_dir:
+                keyresult = self.bucket.list(prefix=rel_path)
+                if len(keyresult) > 0:
+                    exists = True
+                else:
+                    exists = False
+            else:
+                exists = True if self.bucket.get(rel_path) is not None else False
+        except Exception:
+            log.exception("Trouble checking existence of S3 key '%s'", rel_path)
+            return False
+        if rel_path[0] == '/':
+            raise
+        return exists
+
+    def _in_cache(self, rel_path):
+        """ Check if the given dataset is in the local cache and return True if so. """
+        # log.debug("------ Checking cache for rel_path %s" % rel_path)
+        cache_path = self._get_cache_path(rel_path)
+        return os.path.exists(cache_path)
+
+    def _pull_into_cache(self, rel_path):
+        # Ensure the cache directory structure exists (e.g., dataset_#_files/)
+        rel_path_dir = os.path.dirname(rel_path)
+        if not os.path.exists(self._get_cache_path(rel_path_dir)):
+            os.makedirs(self._get_cache_path(rel_path_dir))
+        # Now pull in the file
+        file_ok = self._download(rel_path)
+        self._fix_permissions(self._get_cache_path(rel_path_dir))
+        return file_ok
+
+    def _transfer_cb(self, complete, total):
+        self.transfer_progress += 10
+
+    def _download(self, rel_path):
+        try:
+            log.debug("Pulling key '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+            key = self.bucket.get(rel_path)
+            # Test if cache is large enough to hold the new file
+            if self.cache_size > 0 and key.size > self.cache_size:
+                log.critical("File %s is larger (%s) than the cache size (%s). Cannot download.",
+                             rel_path, key.size, self.cache_size)
+                return False
+            if self.use_axel:
+                log.debug("Parallel pulled key '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+                ncores = multiprocessing.cpu_count()
+                url = key.generate_url(7200)
+                ret_code = subprocess.call("axel -a -n %s '%s'" % (ncores, url))
```

Comment:
@VJalili @afgane This axel based downloading looks interesting - I wonder whether we should consider integrating similar functionality into cloudbridge? Not necessarily through axel, but at least a better implementation than the single threaded one we have now - maybe using an asyncio loop?